lib/rucio/core/transfer.py

```diff
@@ -1382,14 +1398,25 @@ def __list_transfer_requests_and_source_replicas(total_workers=0, worker_number=
             .filter(models.RSEAttrAssociation.key == 'transfertool',
                     models.RSEAttrAssociation.value.like('%' + transfertool + '%'))
 
-    if rses:
-        result = []
-        for item in query.all():
-            dest_rse_id = item[10]
-            if dest_rse_id in rses:
-                result.append(item)
-        return result
-    return query.all()
+    requests_by_id = {}
+    for (request_id, rule_id, scope, name, md5, adler32, byte_count, activity, attributes, previous_attempt_id, dest_rse_id, account, retry_count,
+         source_rse_id, source_rse_name, file_path, source_ranking, distance_ranking) in query:
+
+        # rses (of unknown length) should be a temporary table to check against instead of this special case
+        if rses and dest_rse_id not in rses:
+            continue
+
+        request = requests_by_id.get(request_id)
+        if not request:
+            request = RequestWithSources(id=request_id, rule_id=rule_id, scope=scope, name=name, md5=md5, adler32=adler32, byte_count=byte_count,
+                                         activity=activity, attributes=attributes, previous_attempt_id=previous_attempt_id, dest_rse_id=dest_rse_id,
+                                         account=account, retry_count=retry_count)
```

Comment:
is there any specific reason not to use a `namedtuple` and get rid of great unwrapping in the for-loop?

Reply:
Namedtuples are immutable. There were a couple of places where having a fully mutable class allowed to simplify the logic. For example: to sanitize the source_ranking and set it to 0 in the case when it is returned as None. I'm however not very sure the benefit is really justified right now. Named tuples could do the trick quite well. It ads complexity in some places, but makes life easier in others.... However, I'm thinking of (maybe) bringing some logic into this class in the future for a more object-oriented design. But I'm far from that design for now, and I'll maybe revert back to named tuples if I find it better suited latter.